In [16]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# **STEP 1: Read and view file**

In [17]:
API_edu = pd.read_csv(r"/content/drive/My Drive/0. STUDYING Peo Peo/PYTHON/xAPI-Edu-Data.csv")
API_edu.head()

gender NationalITy  ... StudentAbsenceDays Class
0      M          KW  ...            Under-7     M
1      M          KW  ...            Under-7     M
2      M          KW  ...            Above-7     L
3      M          KW  ...            Above-7     L
4      M          KW  ...            Above-7     M

[5 rows x 17 columns]

# **STEP 2: Kiểm tra dữ liệu khuyết**

In [18]:
API_edu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 480 entries, 0 to 479
Data columns (total 17 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   gender                    480 non-null    object
 1   NationalITy               480 non-null    object
 2   PlaceofBirth              480 non-null    object
 3   StageID                   480 non-null    object
 4   GradeID                   480 non-null    object
 5   SectionID                 480 non-null    object
 6   Topic                     480 non-null    object
 7   Semester                  480 non-null    object
 8   Relation                  480 non-null    object
 9   raisedhands               480 non-null    int64 
 10  VisITedResources          480 non-null    int64 
 11  AnnouncementsView         480 non-null    int64 
 12  Discussion                480 non-null    int64 
 13  ParentAnsweringSurvey     480 non-null    object
 14  ParentschoolSatisfaction  

# **STEP 3**: 
##**1. Tách X va y (dataframe)**

In [19]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()

# Tạo biến target
y = API_edu['Class']
print(y)
print(y.unique())

#  Tập X được chia thành 2 loại:
## Data có dữ liệu là continuous (Dữ liệu chứa số)
X_continuous = API_edu[['Discussion', 'AnnouncementsView', 'VisITedResources', 'raisedhands']]
## Data với dữ liệu là category (Dữ liệu chứa chữ)
X_category = API_edu.drop(['Class', 'Discussion', 'AnnouncementsView', 'VisITedResources', 'raisedhands'], axis = 1)
## Encoding các data
X_category_encode = pd.get_dummies(X_category, prefix_sep='_')
y = le.fit_transform(y)
X = pd.concat([X_continuous, X_category_encode], axis = 1) 

print(y)


0      M
1      M
2      L
3      L
4      M
      ..
475    L
476    M
477    M
478    L
479    L
Name: Class, Length: 480, dtype: object
['M' 'L' 'H']
[2 2 1 1 2 2 1 2 2 2 0 2 1 1 0 2 2 2 2 0 2 2 2 1 1 1 2 1 2 2 0 1 1 1 1 1 1
 2 1 2 1 2 1 2 2 1 1 0 0 2 1 1 2 0 1 1 1 1 2 2 1 2 0 2 1 1 2 0 0 2 1 2 1 2
 1 2 2 2 1 0 1 1 1 2 0 1 0 1 1 1 1 0 0 0 1 0 0 2 2 2 2 0 1 1 2 1 2 0 2 2 0
 2 1 1 1 1 2 0 2 2 2 1 2 2 1 1 2 1 1 1 1 2 1 1 0 0 0 2 0 2 1 1 2 0 1 2 0 2
 2 0 0 2 0 1 2 0 2 2 1 2 0 2 0 2 2 0 2 0 0 2 0 2 1 1 2 1 0 2 0 2 0 1 0 2 1
 0 2 2 0 2 1 1 2 2 2 2 0 0 1 2 0 0 2 2 1 0 2 2 2 2 0 2 0 1 1 1 2 2 0 2 2 2
 2 0 0 2 1 1 0 1 2 1 2 2 2 1 1 2 2 0 0 2 1 2 0 2 0 2 1 2 0 1 2 1 0 0 0 2 2
 1 1 2 2 2 2 0 2 2 2 2 0 2 2 0 0 0 0 0 2 2 0 0 0 0 2 2 0 0 2 2 1 1 0 0 2 2
 0 0 2 2 1 1 2 2 2 2 0 0 2 2 2 2 0 0 0 0 0 0 0 0 2 2 1 1 2 2 1 1 2 2 1 1 2
 2 1 1 2 2 2 2 2 2 2 2 0 0 0 0 1 1 1 1 2 2 0 0 2 2 0 0 2 2 0 0 0 0 2 2 0 0
 2 2 1 1 1 1 2 2 1 1 1 1 0 0 0 0 2 2 1 1 2 2 0 0 0 0 2 2 0 0 2 2 0 0 0 0 1
 1 2 2 0 0 2 2 1 1 0 0

**Comments**

See y after encoding:

M = 2

L = 1

H = 0

##**2. Train_Test**

In [20]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, 
                                                    stratify = y, random_state = 0)

#**STEP 4: Choose model**

##**1. Model: Support Vector Machine (SVM)**

In [21]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix,classification_report

clf_svm = SVC(kernel = 'poly', C = 10)   
clf_svm.fit(X_train, y_train)

print("Support Vector Machine")
print("\ntrain acc: ", accuracy_score(y_train, clf_svm.predict(X_train)))
print("test acc: ", accuracy_score(y_test, clf_svm.predict(X_test)))
print('\nConfusion matrix : \n', confusion_matrix(y_test, clf_svm.predict(X_test)))
print('Classification report : \n', classification_report(y_test, clf_svm.predict(X_test), digits = 5))

Support Vector Machine

train acc:  0.7202380952380952
test acc:  0.6388888888888888

Confusion matrix : 
 [[19  3 21]
 [ 0 34  4]
 [ 9 15 39]]
Classification report : 
               precision    recall  f1-score   support

           0    0.67857   0.44186   0.53521        43
           1    0.65385   0.89474   0.75556        38
           2    0.60938   0.61905   0.61417        63

    accuracy                        0.63889       144
   macro avg    0.64726   0.65188   0.63498       144
weighted avg    0.64177   0.63889   0.62790       144



**Comments:**

Từ confusion matrix chỉ ra rằng trong tổng số 144 dữ liệu của test-set thì có:

 [[19  3 21]

  [ 0 34  4]

  [ 9 15 39]]

* Dòng 0 (H): High-Level Students
    - Thực tế: 43 students
    - (0,0): predict 19 High-Level students => dự đoán ĐÚNG
    - (0,1): predict 3 Low-Level students => dự đoán sai
    - (0,2): predict 21 Middle-Level students => dự đoán sai
* Dòng 1 (L): Low-Level Students
    - Thực tế: 38 students
    - (0,0): predict 0 High-Level students 
    - (0,1): predict 34 Low-Level students => dự đoán ĐÚNG
    - (0,2): predict 4 Middle-Level students => dự đoán sai
* Dòng 2 (M): Middle-Level Students
    - Thực tế: 63 students
    - (0,0): predict 9 High-Level students => dự đoán sai
    - (0,1): predict 15 Low-Level students => dự đoán sai
    - (0,2): predict 39 Middle-Level students => dự đoán ĐÚNG

Bài toán: accuracy  = 63,889%; 


##**2. Model: Decision Tree Classifier**


In [22]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix,classification_report

clf = DecisionTreeClassifier()
clf.fit(X_train,y_train)

print("Decision Tree Classifier")
print("\ntrain acc: ", accuracy_score(y_train, clf.predict(X_train)))
print("test acc: ", accuracy_score(y_test, clf.predict(X_test)))
print('\nConfusion matrix : \n', confusion_matrix(y_test, clf.predict(X_test)))
print('Classification report : \n', classification_report(y_test, clf.predict(X_test), digits = 5))

Decision Tree Classifier

train acc:  1.0
test acc:  0.7708333333333334

Confusion matrix : 
 [[32  0 11]
 [ 0 34  4]
 [10  8 45]]
Classification report : 
               precision    recall  f1-score   support

           0    0.76190   0.74419   0.75294        43
           1    0.80952   0.89474   0.85000        38
           2    0.75000   0.71429   0.73171        63

    accuracy                        0.77083       144
   macro avg    0.77381   0.78440   0.77822       144
weighted avg    0.76926   0.77083   0.76926       144



**Comments:**

Từ confusion matrix chỉ ra rằng trong tổng số 144 dữ liệu của test-set thì có:

[[32  0 11]

 [ 0 34  4]

 [10  8 45]]

* Dòng 0 (H): High-Level Students
    - Thực tế: 43 students
    - (0,0): predict 32 High-Level students => dự đoán ĐÚNG
    - (0,1): predict 0 Low-Level students => dự đoán sai
    - (0,2): predict 11 Middle-Level students => dự đoán sai
* Dòng 1 (L): Low-Level Students
    - Thực tế: 38 students
    - (0,0): predict 0 High-Level students 
    - (0,1): predict 34 Low-Level students => dự đoán ĐÚNG
    - (0,2): predict 4 Middle-Level students => dự đoán sai
* Dòng 2 (M): Middle-Level Students
    - Thực tế: 63 students
    - (0,0): predict 10 High-Level students => dự đoán sai
    - (0,1): predict 8 Low-Level students => dự đoán sai
    - (0,2): predict 45 Middle-Level students => dự đoán ĐÚNG

Bài toán: accuracy  = 77,083%; 


##**3. Model: Random Forest Classifier**

In [23]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix,classification_report

clf = RandomForestClassifier()
clf.fit(X_train,y_train)

print("Random Forest Classifier")
print("\ntrain acc: ", accuracy_score(y_train, clf.predict(X_train)))
print("test acc: ", accuracy_score(y_test, clf.predict(X_test)))
print('\nConfusion matrix : \n', confusion_matrix(y_test, clf.predict(X_test)))
print('Classification report : \n', classification_report(y_test, clf.predict(X_test), digits = 5))

Random Forest Classifier

train acc:  1.0
test acc:  0.8263888888888888

Confusion matrix : 
 [[33  0 10]
 [ 0 34  4]
 [ 5  6 52]]
Classification report : 
               precision    recall  f1-score   support

           0    0.86842   0.76744   0.81481        43
           1    0.85000   0.89474   0.87179        38
           2    0.78788   0.82540   0.80620        63

    accuracy                        0.82639       144
   macro avg    0.83543   0.82919   0.83094       144
weighted avg    0.82832   0.82639   0.82608       144



**Comments:**

Từ confusion matrix chỉ ra rằng trong tổng số 144 dữ liệu của test-set thì có:

 [[33  0 10]

 [ 0 34  4]

 [ 5  6 52]]

* Dòng 0 (H): High-Level Students
    - Thực tế: 43 students
    - (0,0): predict 33 High-Level students => dự đoán ĐÚNG
    - (0,1): predict 0 Low-Level students => dự đoán sai
    - (0,2): predict 10 Middle-Level students => dự đoán sai
* Dòng 1 (L): Low-Level Students
    - Thực tế: 38 students
    - (0,0): predict 0 High-Level students 
    - (0,1): predict 34 Low-Level students => dự đoán ĐÚNG
    - (0,2): predict 4 Middle-Level students => dự đoán sai
* Dòng 2 (M): Middle-Level Students
    - Thực tế: 63 students
    - (0,0): predict 5 High-Level students => dự đoán sai
    - (0,1): predict 6 Low-Level students => dự đoán sai
    - (0,2): predict 52 Middle-Level students => dự đoán ĐÚNG

Bài toán: accuracy  = 82,639%; 
